src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java

```diff
@@ -0,0 +1,116 @@
+    private int minimumRequiredReadsPerStrand = DEFAULT_MINIMUM_READS_PER_STRAND;
+
+    protected ReadsWithSameUMI currentReadsWithSameUMI = null;
+
+    /***
+     * FGBio GroupByUMI returns reads sorted by molecule ID: For example, the input bam may look like
+     * read1: ... MI:Z:0/A ...
+     * read2: ... MI:Z:0/A ...
+     * read3: ... MI:Z:0/B ...
+     * read4: ... MI:Z:0/B ...
+     * read5: ... MI:Z:1/A ...
+     * read6: ... MI:Z:1/B ...
+     * read7: ... MI:Z:1/B ...
+     *
+     * Thus it's sufficient to go through the reads in order and collect them in a list until
+     * we encounter the next molecule ID, at which point we pass the list to the {@code apply} method,
+     * process the set based on the child class's implementation of the method, and clear the {@code currentDuplicateSet} variable and start collecting reads again.
+     *
+     * Notice there are two apply() methods in this class:
+     * This apply() inherited from ReadWalker is marked final to discourage subclassing.
+     * A subclass must override the other apply() method that takes in the DuplicateSet.
+     */
+    @Override
+    public final void apply(GATKRead read, ReferenceContext referenceContext, FeatureContext featureContext) {
+        if (currentReadsWithSameUMI == null){ // evaluates to true for the very first read
+            currentReadsWithSameUMI = new ReadsWithSameUMI(read);
+            return;
+        }
+
+        final int readMoleculeNumber = MoleculeID.getMoleculeNumberOfRead(read);
+        final int duplicateSetMoleculeNumber = currentReadsWithSameUMI.getMoleculeNumber();
+
+        // If the incoming read has the molecule id less than that of the currentDuplicateSet,
+        // the input bam is not sorted properly by the MI tag
+        if (duplicateSetMoleculeNumber > readMoleculeNumber){
+            throw new UserException(String.format("The input bam must be sorted by the molecule ID (%s) tag.", SAMTag.MI.name()));
+        }
+
+        if (duplicateSetMoleculeNumber < readMoleculeNumber) {
+            // The incoming read's molecule ID does not match that of the current duplicate, meaning we've reached the end of the current set.
+            // Call the apply() method to process the current set and start a new set.
+            if (rejectSet(currentReadsWithSameUMI)){
+                currentReadsWithSameUMI = new ReadsWithSameUMI(read);
+                return;
+            }
+
+            apply(currentReadsWithSameUMI,
+                    new ReferenceContext(reference, currentReadsWithSameUMI.getInterval()), // Will create an empty ReferenceContext if reference or readInterval == null
+                    new FeatureContext(features, currentReadsWithSameUMI.getInterval()));
+            currentReadsWithSameUMI = new ReadsWithSameUMI(read);
+            return;
+        }
+
+        // The incoming read has the same UMI as the current set; simply add to the current set
```

Comment:
I like to put the exclusive conditions in an else block to make it clear that none of the options above fall down into this, but that's a style point that some people disagree with.

Reply:
Yea I didn't like the way the if else blocks looked so I did it this way, hope that's ok.